### ISSUE TITLE
Closing menu does not restore focus of a widget

### ISSUE THREAD

**Issue by beru** [cnjinhao/nana#37]
Below is the code and steps to reproduce the issue.
```c++
#include <nana/gui.hpp>
#include <nana/gui/widgets/menubar.hpp>
#include <nana/gui/widgets/textbox.hpp>

using namespace nana;

int main()
{
	using namespace nana;
	form fm(API::make_center(400, 500));
	menubar mb;
	mb.create(fm);
	mb.push_back(L"&aaaaaa");
	mb.push_back(L"&baaaaa");
	mb.at(0).append(L"&1", [](menu::item_proxy& ip){
	
	});
	mb.at(0).append(L"&2", [](menu::item_proxy& ip){
	
	});
	mb.at(1).append(L"&1", [](menu::item_proxy& ip){
	
	});
	mb.at(1).append(L"&2", [](menu::item_proxy& ip){
	
	});

	textbox txt(fm, rectangle(10, 60, 200, 30));

	fm.show();
	exec();
}
```
1. Click and set focus to the textbox.
2. Click the menubar to open the menu.
3. Click a menuitem.
4. Focus of the textbox widget is lost.

**Comment by beru** [comment]
I tried to work around the issue by modifying nana's source code.
File : **nana\source\gui\detail\window_manager.cpp**
Func : **nana::detail::window_manager::set_focus** bottom part.
```c++
#if 1
				if (wd == wd->root_widget->other.attribute.root->menubar) {
					brock.set_menubar_taken(prev_focus);
				}else {
					brock.set_menubar_taken(wd);
				}
#else
				brock.set_menubar_taken(wd);
#endif
			}
```

File : **nana\source\gui\detail\win32\bedrock.cpp**
Func : **Bedrock_WIN32_WindowProc**
When WM_SYSKEYDOWN comes, window_manager::set_focus is called with the menubar window.

And in
File : **nana\source\gui\widgets\menubar.cpp**
Func : **nana::drawerbase::menubar::trigger::_m_total_close**
API::restore_menubar_taken_window is called.

The idea is restoring menubar's focus when the menubar closes is odd and meaningless,
so I thought calling **set_menubar_taken** with prev_focus window would be a good alternative.

But there's a problem with this workaround.
After pressing ALT + some key to select a menu item, WM_KEYDOWN -> WM_CHAR will be sent and focus restored window (widget) has to handle the key. In case textbox is prev-focus widget, menu item's shortcut key character goes to the textbox against my wish.

To fix the problem, it seems like **TranslateAccelerator** API should be called before **TranslateMessage** API in a message loop so that accelerators and shortcut keys can survive.

**Comment by beru** [comment]
It seems like standard Windows applications use CreateAcceleratorTable & TranslateAccelerator APIs.
nana does not rely on those APIs to handle user inputs.
I really feel GUI message handling is deep... it is not trivial.

**Comment by cnjinhao** [comment]
Thanks for your information, a fix has been checked into the hotfixes branch.

**Comment by beru** [comment]
Thank you for the fix!

I've tested and noticed there still remains a lot of tiny issues regarding this restore-focus-when-a-menu-closes feature.
such as....
Closing a menu item by pressing enter key restores the original focus. **OK**
but,
Closing a menu item with shortcut key (a letter after ampersand in a caption string) does not restore the original focus. **NG**
Clicking a menu item does not restore the original focus. **NG**
Pressing **alt** key to cancel menubar focus does not restore the original focus. **NG**

Clicking a title bar does not restore the original focus. **NG**
Belows are some steps to reproduce the issue.
1. Click a textbox. (focus of this widget item should be restored later)
2. Click a menu on a menubar.
3. Click title bar of test application window.
4. the menu closes but the focus is not restored.

To resolve the issue, I could think of below modification.
in file : source\gui\detail\win32\bedrock.cpp
in func : nana::detail::Bedrock_WIN32_WindowProc
```c++
case WM_NCLBUTTONDOWN: case WM_NCMBUTTONDOWN: case WM_NCRBUTTONDOWN:
				brock.close_menu_if_focus_other_window(native_window);
				def_window_proc = true;
				// ad hoc workaround...
				API::restore_menubar_taken_window();
				break;
```

And current implmentation does not restore focus of a widget control which was focused before opening a menu when application's focus is lost and the menu closes. **NG**
Below are the modification I tested without understanding many logics in library code.
in file : source\gui\detail\win32\bedrock.cpp
in func : nana::detail::Bedrock_WIN32_WindowProc
```c++
			case WM_KILLFOCUS:
				if(msgwnd->other.attribute.root->focus)
				{
					auto focus = msgwnd->other.attribute.root->focus;
					if (focus != msgwnd->root_widget->other.attribute.root->menubar)
					{
						arg_focus arg;
						assign_arg(arg, focus, reinterpret_cast<native_window_type>(wParam), false);
						if(brock.emit(event_code::focus, focus, arg, true, &context))
						{
							if(focus->together.caret)
								focus->together.caret->set_active(false);
						}
					}
					//wParam indicates a handle of window that receives the focus.
					brock.close_menu_if_focus_other_window(reinterpret_cast<native_window_type>(wParam));
					API::restore_menubar_taken_window();
				}
				//focus_changed means that during an event procedure if the focus is changed
				if(brock.wd_manager.available(msgwnd))
					msgwnd->root_widget->other.attribute.root->context.focus_changed = true;
				break;
```

Frankly saying, I don't think putting workarounds into bedrock.cpp is favourable because doing so can blur structures and logics of the library.

There're still more and more tiny behaviour differences with nana's menu and standard menu.
so fixing those can be a long journey.

**Comment by cnjinhao** [comment]
Very useful feedback! only one issue left. "Pressing alt key to cancel menubar focus does not restore the original focus. NG"

**Comment by beru** [comment]
My previous post was not well organized. Let me list current status of the issue.
- [x] Closing a menu item by pressing enter key restores the original focus. (it's already implemented)
- [ ] Closing a menu item with shortcut key (a letter after ampersand in a caption string) does not restore the original focus. (not yet fixed)
- [ ] Clicking a menu item does not restore the original focus. (not yet fixed)
- [ ] Pressing alt key to cancel menubar focus does not restore the original focus.  (not yet fixed)
- [x] Clicking a title bar does not restore the original focus. (there's a workaround in the above post)
- [x] When application's focus is lost and its menu closes, focus of a widget control which was focused before opening a menu is not restored. (there's a workaround in the above post)

**Comment by cnjinhao** [comment]
Thank you! a fix has been checked into the hotfixes branch.

**Comment by beru** [comment]
Thank you for the fix.
I have pulled the hotfixes branch and confirmed that many reported issues are now fixed.
But unfortunately the fix has created several new issues and there are numerous remaining issues...

Let me write some of those.

From what I observe from other Windows applications such as Explorer and Notepad, pressing ESC key while a menu is opened should close the menu but the focus of the menubar should not be lost. Pressing ESC key **once again** should remove the focus from the menubar.

But currently in the nana hotfixes branch build, pressing ESC key closes opening menu and the focus of the menubar is lost **together at once**. I think this behaviour is unwanted since users tend to cancel menu selection by pressing ESC key and press other menu shortcut keys for subsequent operations.

And I noticed there is another issue. After I've opened a menu with shortcut key (a letter after ampersand letter in a menu caption), It's not always but sometimes (about 30%? YMMV)
the menubar does not handle key inputs. so I have to use my (device) mouse to use an application.

**Comment by cnjinhao** [comment]
Thanks for your information, and a fix has been checked into the hotfixes branch.

**Comment by beru** [comment]
Thanks you for the fix. I noticed there's a problem.

```c++
#include <nana/gui.hpp>
#include <nana/gui/widgets/menubar.hpp>

using namespace nana;

int main()
{
	using namespace nana;
	form fm(API::make_center((unsigned int)(400*1.414), 400));
	menubar mb;
	mb.create(fm);
	mb.push_back(L"&File");
	auto& file_menu = mb.at(0);
	file_menu.append(L"&Open", [&](menu::item_proxy& ip){
	});
	file_menu.append(L"&Save", [&](menu::item_proxy& ip){
	});
	fm.modality();
}
```

1. Keydown ALT key.
2. Keydown F key.
 - File menu pops up on the menubar.
3. Keyup F key.
 -  File menu closes but this is **not** unwanted.

**Comment by cnjinhao** [comment]
A fix has been shown up in release 1.0.2

**Comment by beru** [comment]
Thank you for the fix!
